ff/serving/provider/postgres.go

```diff
@@ -0,0 +1,501 @@
+		return err
+	}
+	if exists, err := table.resourceExists(rec); err != nil {
+		return err
+	} else if !exists {
+		query := fmt.Sprintf("INSERT INTO %s (entity, value, ts) VALUES ($1, $2, $3) RETURNING entity, value, ts", tb)
+		if _, err := table.conn.Exec(context.Background(), query, rec.Entity, value, rec.TS); err != nil {
+			return err
+		}
+	} else if exists {
+		query := fmt.Sprintf("UPDATE %s SET value=$3 WHERE Entity=$1 AND ts=$2 RETURNING entity, value, ts", tb)
+		if _, err := table.conn.Exec(context.Background(), query, rec.Entity, rec.TS, value); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (table *postgresOfflineTable) resourceExists(rec ResourceRecord) (bool, error) {
+	query := fmt.Sprintf("SELECT entity, value, ts FROM %s WHERE entity=$1 AND ts=$2 ", sanitize(table.name))
+	rows, err := table.conn.Query(context.Background(), query, rec.Entity, rec.TS)
+	defer rows.Close()
+	if err != nil {
+		return false, err
+	}
+	rowCount := 0
+	for rows.Next() {
+		rowCount++
+	}
+	if rowCount == 0 {
+		return false, nil
+	}
+	return true, nil
+}
+
+type postgresMaterialization struct {
+	id        MaterializationID
+	conn      *pgxpool.Pool
+	ctx       context.Context
+	tableName string
+	data      []ResourceRecord
+}
+
+func (mat *postgresMaterialization) ID() MaterializationID {
+	return mat.id
+}
+
+func (mat *postgresMaterialization) NumRows() (int64, error) {
+	n := int64(0)
+	query := fmt.Sprintf("SELECT COUNT(*) FROM %s", sanitize(mat.tableName))
+	rows := mat.conn.QueryRow(context.Background(), query)
+	err := rows.Scan(&n)
+	if err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+func (mat *postgresMaterialization) deserialize(v []byte) (postgresTableItem, error) {
+	item := postgresTableItem{}
+	if err := json.Unmarshal(v, &item); err != nil {
+		return postgresTableItem{}, err
+	}
+	return item, nil
+}
+
+// castTableItemType returns the value casted as its original type
+func castTableItemType(v postgresTableItem) interface{} {
+	switch v.ItemType {
+	case "int":
+		return int(v.Value.(float64))
+	case "int8":
+		return int8(v.Value.(float64))
+	case "int16":
+		return int16(v.Value.(float64))
+	case "int32":
+		return int32(v.Value.(float64))
+	case "int64":
+		return int64(v.Value.(float64))
+	case "float32":
+		return float32(v.Value.(float64))
+	case "float64":
+		return v.Value
+	case "string":
+		return v.Value.(string)
+	case "bool":
+		return v.Value.(bool)
+	default:
+		return v.Value
+	}
+}
+
+func (mat *postgresMaterialization) IterateSegment(start, end int64) (FeatureIterator, error) {
+	query := fmt.Sprintf(""+
+		"SELECT entity, value, ts::timestamptz FROM "+
+		"( SELECT * FROM "+
+		"( SELECT *, row_number() over() FROM %s )t1 WHERE row_number>$1 AND row_number<=$2)t2", sanitize(mat.tableName))
+	rows, err := mat.conn.Query(context.Background(), query, start, end)
+	defer rows.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	data := make([]ResourceRecord, 0)
+	for rows.Next() {
```

Comment:
Let's wrap this in a FeatureIterator?

Reply:
You mean having it get the rows individually from the db instead of requesting them in bulk and placing into memory?

Reply:
Wrapped it up